crates/proto/src/rr/dnssec/tbs.rs

```diff
@@ -7,19 +7,228 @@
+impl TBS {
+    /// Returns the to-be-signed serialization of the given record set using the information
+    /// provided from the RRSIG record.
+    ///
+    /// # Arguments
+    ///
+    /// * `rrsig` - SIG or RRSIG record, which was produced from the RRSet
+    /// * `records` - RRSet records to sign with the information in the `rrsig`
+    ///
+    /// # Return
+    ///
+    /// binary hash of the RRSet with the information from the RRSIG record
+    pub fn from_rrsig<'a>(
+        rrsig: &Record<RRSIG>,
+        records: impl Iterator<Item = &'a Record>,
+    ) -> ProtoResult<Self> {
+        Self::from_sig(rrsig.name(), rrsig.dns_class(), rrsig.data(), records)
+    }
+
+    /// Returns the to-be-signed serialization of the given record set using the information
+    /// provided from the SIG record.
+    ///
+    /// # Arguments
+    ///
+    /// * `name` - labels of the record to sign
+    /// * `dns_class` - DNSClass of the RRSet, i.e. IN
+    /// * `sig` - SIG or RRSIG record, which was produced from the RRSet
+    /// * `records` - RRSet records to sign with the information in the `rrsig`
+    ///
+    /// # Return
+    ///
+    /// binary hash of the RRSet with the information from the RRSIG record
+    pub fn from_sig<'a>(
+        name: &Name,
+        dns_class: DNSClass,
+        sig: &SIG,
+        records: impl Iterator<Item = &'a Record>,
+    ) -> ProtoResult<Self> {
+        Self::new(
+            name,
+            dns_class,
+            sig.num_labels(),
+            sig.type_covered(),
+            sig.algorithm(),
+            sig.original_ttl(),
+            sig.sig_expiration(),
+            sig.sig_inception(),
+            sig.key_tag(),
+            sig.signer_name(),
+            records,
+        )
+    }
+
+    /// Returns the to-be-signed serialization of the given record set.
+    ///
+    /// # Arguments
+    ///
+    /// * `inception` - the epoch seconds of when this hashed signature will be valid
+    /// * `rr_set` - RRSet to sign
+    /// * `zone_class` - DNSClass, i.e. IN, of the records
+    /// * `signer` - the signer to use for signing the RRSet
+    ///
+    /// # Returns
+    ///
+    /// the binary hash of the specified RRSet and associated information
+    // FIXME: OMG, there are a ton of asserts in here...
+    pub fn from_rrset(
+        rr_set: &RecordSet,
+        zone_class: DNSClass,
+        inception: OffsetDateTime,
+        expiration: OffsetDateTime,
+        signer: &SigSigner,
+    ) -> ProtoResult<Self> {
+        Self::new(
+            rr_set.name(),
+            zone_class,
+            rr_set.name().num_labels(),
+            rr_set.record_type(),
+            signer.algorithm(),
+            rr_set.ttl(),
+            SerialNumber(expiration.unix_timestamp() as u32),
+            SerialNumber(inception.unix_timestamp() as u32),
+            signer.calculate_key_tag()?,
+            signer.signer_name(),
+            // TODO: this is a nasty clone... the issue is that the vec
```

Comment:
Hm, is the comment referencing the function `records_without_rrsigs` or was there a `clone` call here previously? To me it looks like the function wraps an `Iterator` over `&Record`s.

Reply:
Removed in the commit that fixed the cloning.